test/gtest/common/mem_buffer.h

```diff
@@ -129,6 +130,7 @@ class mem_buffer {
                                      size_t length, size_t offset,
                                      const void *buffer, const void *orig_ptr)
     {
+        ucs_assertv(length <= 8, "length=%zu", length);
```

Comment:
why?

Reply:
Since new implementation of UCS_MASK will still do undefined behavior if input value > 64. In that particular case length is size of one element that should be compared.

Reply:
maybe we can define UCS_MASK the same was UCS_MASK_SAFE is defined and then no need for this..

Reply:
can we make UCS_MASK always 64-bit, and if the result is casted to smaller type it will still be ok?

Reply:
> can we make UCS_MASK always 64-bit, and if the result is casted to smaller type it will still be ok?

Can you show example of the compilation error?

Reply:
That's example of compilation error on Ubuntu 20.04 for macro implementation with `UINT64_MAX`:
```
/__w/2/s/contrib/../src/uct/ib/mlx5/ib_mlx5.c:218:26: error: implicit conversion from 'unsigned long' to 'unsigned int' changes value from 18446744073709551615 to 4294967295 [-Werror,-Wconstant-conversion]
    cq->cq_length_mask = UCS_MASK(cq->cq_length_log);
                       ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/__w/2/s/contrib/../src/ucs/sys/compiler_def.h:83:38: note: expanded from macro 'UCS_MASK'
#define UCS_MASK(_i) (((_i) >= 64) ? 0xFFFFFFFFFFFFFFFF : (UCS_BIT(_i) - 1))
                                    ^~~~~~~~~~~~~~~~~~
```

I will try to implement it through inline function and see whether error would happen then.

But anyway, from my point of view we are walking on a very thin ice here, so even if it would work in our CI, it can break something on some tricky platform. So I would prefer to use different impementations for `UCS_MASK` and `UCS_MASK64`

Reply:
would using `#define UCS_MASK(_i) (((_i) >= 64) ? ~0 : (UCS_BIT(_i) - 1))` fix all cases?

Reply:
@tvegas1 thanks for the suggestion!

I checked C standard and that what does it say about conditional operator:

> If both the second and third operands have arithmetic type, the result type that would be
determined by the usual arithmetic conversions, were they applied to those two operands,
is the type of the result. If both the operands have structure or union type, the result has
that type. If both operands have void type, the result has void type.

So if I understand that correctly, it means that ~0 will be casted to the type that is used as a left operand of assignment.